src/messages/MOSDRepOp.h

```diff
@@ -64,10 +70,40 @@ class MOSDRepOp : public Message {
   }
 
   virtual void decode_payload() {
-    bufferlist::iterator p = payload.begin();
-    ::decode(map_epoch, p);
-    ::decode(reqid, p);
-    ::decode(pgid, p);
+    p = payload.begin();
+    if (header.version < 2) {
+      ::decode(map_epoch, p);
+      ::decode(reqid, p);
+      ::decode(pgid, p);
+      ::decode(poid, p);
+
+      ::decode(acks_wanted, p);
+      ::decode(version, p);
+      ::decode(logbl, p);
+      ::decode(pg_stats, p);
+      ::decode(pg_trim_to, p);
+
+
+      ::decode(new_temp_oid, p);
+      ::decode(discard_temp_oid, p);
+
+      ::decode(from, p);
+      ::decode(updated_hit_set_history, p);
+      ::decode(pg_trim_rollback_to, p);
+    } else {
+      // new, v2 decode, splitted to partial and final
+      ::decode(map_epoch, p);
+      ::decode(reqid, p);
+      ::decode(pgid, p);
```

Comment:
It looks like the encoding format isn't changing at all here, so we don't have to touch the version.  We just need to break the decode into two stages.  But we can do that unconditionally....